Envoy AI Gateway integration

sivanantha321 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. Before raising a PR, please run `make go-lint` and `make py-fmt` to check the code style.
3. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
4. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This diagram illustrates the architecture of AI Gateway integration with KServe InferenceService using Gateway API in RawDeployment mode.


```mermaid
graph TB
    subgraph "User Input"
        U[User] --> ISVC_YAML[InferenceService YAML<br/>- AI Gateway annotation set to true<br/>- RawDeployment mode<br/>- TrafficPolicy with RateLimit<br/>- Gateway API enabled]
    end

    subgraph "KServe Control Plane"
        ISVC_YAML --> ISVC_CONTROLLER[InferenceService Controller]
        ISVC_CONTROLLER --> ISVC_CRD[(InferenceService CRD)]
        
        subgraph "AI Gateway Reconcilers"
            ISVC_CONTROLLER --> ASB_RECONCILER[AIServiceBackend Reconciler]
            ISVC_CONTROLLER --> AGR_RECONCILER[AIGatewayRoute Reconciler]
            ISVC_CONTROLLER --> BTP_RECONCILER[BackendTrafficPolicy Reconciler]
        end
    end

    subgraph "Created Resources"
        ISVC_CONTROLLER --> PREDICTOR_DEPLOY[Predictor Deployment<br/>- Hugging Face Model<br/>- VLLM Backend<br/>- OpenAI Compatible API]
        PREDICTOR_DEPLOY --> K8S_SERVICE[Kubernetes Service<br/>- Predictor Service<br/>- ClusterIP/LoadBalancer]
        
        ASB_RECONCILER --> ASB[AIServiceBackend<br/>- Links to K8s Service<br/>- OpenAI Schema Config<br/>- Backend Reference]
        AGR_RECONCILER --> AGR[AIGatewayRoute<br/>- Unified AI API<br/>- Schema Definition<br/>- Route Configuration]
        BTP_RECONCILER --> BTP[BackendTrafficPolicy<br/>- Rate Limiting Rules<br/>- Token-based Tracking]
        
        %% Connect to respective management systems
        ASB --> ASB_MGMT
        AGR --> AGR_MGMT
        BTP --> BTP_MGMT
    end

    subgraph "Gateway Infrastructure"
        subgraph "Envoy AI Gateway"
            EAG[Envoy AI Gateway<br/>- External Processing<br/>- Model Routing<br/>- Content Translation]
            EAG --> ASB_MGMT[AIServiceBackend Management]
            EAG --> AGR_MGMT[AIGatewayRoute Management]
            EAG --> EXT_PROC[External Processor<br/>- AI Gateway Logic<br/>- Model Routing<br/>- Content Transformation]
            EXT_PROC --> EXT_PROC_DEPLOY[External Processor Deployment<br/>- ConfigMap<br/>- Service]
        end
        
        subgraph "Envoy Gateway"
            ENVOY_GATEWAY[Envoy Gateway<br/>- Gateway API Implementation<br/>- Load Balancing<br/>- TLS Termination]
            ENVOY_GATEWAY --> BTP_MGMT[BackendTrafficPolicy Management]
            ENVOY_GATEWAY --> ENVOY_PROXY[Envoy Proxy<br/>- Rate Limiting<br/>- Traffic Management<br/>- Request/Response Forwarding]
        end
        
        %% Integration between Envoy AI Gateway and Envoy Gateway
        EAG -.->|External Processor Integration| ENVOY_GATEWAY
    end

    subgraph "Gateway API Resources"
        AGR --> HTTPROUTE[HTTPRoute<br/>- Gateway Binding<br/>- Traffic Routing<br/>- Path-based Routing]
        HTTPROUTE --> GATEWAY[Gateway<br/>- Entry Point<br/>- Network Configuration]
        GATEWAY --> ENVOY_GATEWAY
    end

    subgraph "Cross-Namespace Access"
        REF_GRANT[ReferenceGrant<br/>- Cross-namespace Access<br/>- Security Policy<br/>- RBAC]
        REF_GRANT -.-> ASB
        REF_GRANT -.-> K8S_SERVICE
    end

    subgraph "Client Interaction"
        CLIENT[AI Client<br/>- OpenAI SDK<br/>- Chat Completions<br/>- Text Generation] --> ENVOY_PROXY
        ENVOY_PROXY --> K8S_SERVICE
        K8S_SERVICE --> PREDICTOR_DEPLOY
    end

    subgraph "Rate Limiting Flow"
        ENVOY_PROXY --> RATE_LIMIT_CHECK{Rate Limit Check<br/>- User ID Header<br/>- Model Name<br/>- Token Count}
        RATE_LIMIT_CHECK -->|Within Limit| PREDICTOR_DEPLOY
        RATE_LIMIT_CHECK -->|Exceeded| RATE_LIMIT_RESPONSE[429 Too Many Requests]
    end

    subgraph "Monitoring & Observability"
        PREDICTOR_DEPLOY --> METRICS[Metrics<br/>- Request Count<br/>- Token Usage<br/>- Response Time]
        ENVOY_PROXY --> TRACING[Distributed Tracing<br/>- Request Flow<br/>- Performance Analysis]
        BTP --> LOGGING[Rate Limit Logs<br/>- Audit Trail<br/>- Policy Enforcement]
    end

    %% Styling
    classDef userInput fill:#e1f5fe
    classDef controller fill:#f3e5f5
    classDef resource fill:#e8f5e8
    classDef gateway fill:#fff3e0
    classDef envoyAI fill:#f1f8e9
    classDef envoyGW fill:#fce4ec
    classDef client fill:#e3f2fd

    class U,ISVC_YAML userInput
    class ISVC_CONTROLLER,ASB_RECONCILER,AGR_RECONCILER,BTP_RECONCILER controller
    class PREDICTOR_DEPLOY,K8S_SERVICE,ASB,AGR,BTP,REF_GRANT resource
    class HTTPROUTE,GATEWAY gateway
    class EAG,ASB_MGMT,AGR_MGMT,EXT_PROC,EXT_PROC_DEPLOY envoyAI
    class ENVOY_GATEWAY,BTP_MGMT,ENVOY_PROXY envoyGW
    class CLIENT,RATE_LIMIT_CHECK,RATE_LIMIT_RESPONSE client
```
## Key Components

### Prerequisites
- **Gateway API**: Must be enabled in the Kserve
- **Envoy Gateway**: Should be installed and configured
- **Envoy AI Gateway**: Should be installed and configured

### 1. **User Input Layer**
- User creates InferenceService with AI Gateway annotation set to true
- Specifies RawDeployment mode and TrafficPolicy with rate limiting
- Enables Gateway API integration

Example InferenceService:
```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: qwen2
  namespace: kserve-test
  annotations:
    serving.kserve.io/deploymentMode: RawDeployment
    serving.kserve.io/enable-aigateway: "true"
spec:
  predictor:
    model:
      modelFormat:
        name: huggingface
      args:
        - --model_name=qwen2
        - --model_id=Qwen/Qwen2-0.5B-Instruct
      env:
        - name: KSERVE_OPENAI_ROUTE_PREFIX # Disable OpenAI Route Prefix
          value: ""
      resources:
        limits:
          cpu: "1"
          memory: 8Gi
        requests:
          cpu: "1"
          memory: 4Gi
  trafficPolicy:
    rateLimit:
      global:
        rules:
          # Rate limit rule for qwen2: 1000 total tokens per hour per user
          - clientSelectors:
              - headers:
                  - name: x-user-id
                    type: Distinct
                  - name: x-ai-eg-model
                    type: Exact
                    value: qwen2
            limit:
              requests: 100    # 1000 total tokens per hour
              unit: Hour
            cost:
              request:
                from: Number
                number: 0      # Set to 0 so only token usage counts
              response:
                from: Metadata
                metadata:
                  namespace: io.envoy.ai_gateway
                  key: llm_total_token    # Uses total tokens from the responses
```

### 2. **KServe Control Plane**
- **InferenceService Controller**: Main orchestrator
- **AIServiceBackend Reconciler**: Creates and manages backend resources
- **AIGatewayRoute Reconciler**: Creates and manages route resources
- **BackendTrafficPolicy Reconciler**: Creates and manages traffic policies

### 3. **Created Resources**
- **Predictor Deployment**: Runs the ML model (Hugging Face + VLLM)
- **Kubernetes Service**: Exposes the predictor internally
- **AIServiceBackend**: Links to K8s service with OpenAI schema config
- **AIGatewayRoute**: Defines unified AI API and routing rules
- **BackendTrafficPolicy**: Applies rate limiting policies
- **HttpRoute**: Created as part of default behaviour. But not used with AI Gateway.

### 4. **Gateway Infrastructure**
- **Envoy AI Gateway**: Handles AI-specific processing, model routing, and content translation
  - Manages AIServiceBackend and AIGatewayRoute resources
  - Provides external processor for AI logic
  - Integrates with Envoy Gateway via external processor
- **Envoy Gateway**: Handles Gateway API implementation and traffic management
  - Manages BackendTrafficPolicy for rate limiting
  - Provides Envoy Proxy for request/response handling
  - Handles TLS termination and load balancing

### 5. **Gateway API Resources**
- **HTTPRoute**: Gateway API routing configuration
- **Gateway**: Entry point with network configuration

### 6. **Security & Access Control**
- **ReferenceGrant**: Enables secure cross-namespace access

## Traffic Flow

1. **Client Request**: AI client sends OpenAI-compatible request
2. **Gateway Entry**: Request enters through Envoy Gateway
3. **Rate Limiting**: Envoy Gateway checks rate limits via BackendTrafficPolicy
4. **AI Processing**: Envoy AI Gateway handles AI-specific transformations via external processor
5. **Model Routing**: Envoy AI Gateway routes to appropriate backend using AIServiceBackend
6. **Backend Routing**: Request routed to predictor service via AIServiceBackend
7. **Model Inference**: Hugging Face model processes the request
8. **Response Transform**: Envoy AI Gateway converts response to OpenAI format
9. **Client Response**: Formatted response returned to client

## Rate Limiting Implementation

- **Header-based**: for example `x-user-id` and `x-ai-eg-model` headers for per user per model rate limiting
- **Token Counting**: Tracks token usage per user per model
- **Time Windows**: Enforces limits per hour/minute/second


Depends on #4321 


**Type of changes**
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Integration with Envoy AI Gateway
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

sivanantha321 committed:
Initial Implementation

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add TrafficPolicy API skeleton to InferenceServiceSpec

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Registre schemes and generate roles

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Use our own RateLimit type

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Implement BackendTrafficPolicy reconciler

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

terrytangyuan reviewed (commented):
Is this still needed given that the documentation has been merged?

johnugeorge commented:
Can we add a flag to enable/disable creation of ai gateway resources even if crds are installed?

sivanantha321 committed:
Bump ai-gateway and keda core

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Rebase

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
remove ClusterSettings from BackendTrafficPolicy creation

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add AIGatewayRoute reconciler and update RBAC permissions for AI Gateway resources

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Refactor InferenceService reconciler to improve ownership tracking and label management

- Updated AIGatewayRouteReconciler to use labels for ownership tracking instead of controller references.
- Enhanced AIServiceBackendReconciler to accept InferenceService as a parameter and updated label handling.
- Introduced a new method to delete AIGatewayRoute and AIServiceBackend resources associated with InferenceService.
- Improved HTTPRoute reconciler to check statuses of all components and set appropriate conditions on InferenceService.
- Refactored BackendTrafficPolicyReconciler to streamline traffic policy creation and deletion.
- Updated agent injector and mutator to use InferenceServiceNameLabel for better consistency in label usage.
- Adjusted unit tests to reflect changes in label keys and ensure proper functionality.

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add unit tests for AIServiceBackend and BackendTrafficPolicy reconciler

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Enhance BackendTrafficPolicyReconciler with error handling and ownership tracking

- Added comments to clarify the handling of controller references in cross-namespace scenarios.
- Introduced ownership labels for BackendTrafficPolicy to track associations with InferenceService.
- Implemented a trafficPolicyClientInterceptor to simulate client errors for testing purposes.
- Added comprehensive test cases for error scenarios in the Reconcile method, including:
  - Handling creation and update errors.
  - Testing dry-run updates.
  - Simulating errors during policy retrieval.
  - Ensuring proper behavior when actual updates fail after successful dry-runs.
- Updated existing tests to use require assertions for better error reporting.

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
lint

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
feat: enhance AI Gateway integration with improved reconcilers and comprehensive testing

This commit enhances the AI Gateway integration for KServe InferenceServices with
several key improvements and comprehensive test coverage.

Key Changes:
- Fix AIGatewayRoute backend weight assignment (uncommented Weight: 1)
- Enhance BackendTrafficPolicy reconciler with conditional rate limiting
- Add comprehensive integration tests for AI Gateway functionality
- Add extensive unit test coverage for reconcilers
- Include CRD definitions for AI Gateway resources

Components Updated:
- AIGatewayRoute reconciler: Fixed backend reference weight configuration
- AIServiceBackend reconciler: Enhanced error handling and functionality
- BackendTrafficPolicy reconciler: Improved rate limiting with null safety
- HTTP route reconciler: Minor cleanup and improvements

Testing:
- Added integration tests covering end-to-end AI Gateway scenarios
- Enhanced unit test coverage for all reconciler components
- Added CRD definitions for test environment setup
- Updated test suite configuration for AI Gateway resources

Technical Details:
- Proper weight assignment for AI Gateway route backend references
- Conditional rate limiting in traffic policies (only when specified)
- Cross-namespace resource management improvements
- Enhanced error handling and logging throughout reconcilers

This completes the AI Gateway integration foundation, enabling advanced
AI-specific routing and traffic management for KServe inference workloads.

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix flaky test

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Fix import alias for gateway API in envtest_setup.go

- Changed the import alias from `gatewayapiv1` to `gwapiv1` for consistency.
- Updated the installation call to use the new alias.

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
add AI Gateway e2e test and setup CI

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix: update logger names in AIServiceBackend and BackendTrafficPolicy reconciler tests
fix: correct file paths in Hugging Face VLLM CPU tests for streaming responses

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
revert:: update import alias for knative apis in httproute reconciler

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix: add network_layer parameter to AI Gateway tests and utility functions

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix: reduce timeout duration in inference service tests and disable cache in k8s client

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix: update RawHTTPRoute reconciler to return ctrl.Result for requeue scenarios and enhance HTTPRoute readiness checks

Fix HTTPRoute reconciler to properly handle requeue scenarios

Updates the RawHTTPRoute reconciler to return ctrl.Result instead of error-only responses, enabling proper requeue handling when HTTPRoutes are not ready or not found.

Enhances HTTPRoute readiness checks by comparing annotations and labels in addition to spec equality, preventing unnecessary updates when routes are semantically equivalent.

Improves status checking logic to requeue reconciliation when HTTPRoutes are not ready rather than silently continuing, ensuring proper convergence to desired state.

Adds comprehensive test coverage for various HTTPRoute readiness scenarios including not found, not ready, and multi-component failure cases.

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
fix: increase sleep duration in explain_response function to address timing issues

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>
